### PR TITLE
Make proposal based on topological torsions

### PR DESCRIPTION
**Pull request by pgrinaway**
Still have a few things to do. I've written code to extract topological torsions from parmed. I need to get these into the proposal/logp bits, as well as allow calculation of improper torsions.

### CODE REVIEW

**Comment by pgrinaway** [comment]
Question: how arbitrary is the ordering of the atoms in a topological torsion (1-2-3-4 vs 4-3-2-1)? I realize this probably shifts the angle, but since the potentials are calculated with their own angle (in the correct order for that torsion), should it make a difference when finding topological torsions?

**Comment by pgrinaway** [comment]
This is probably a good time to move to OpenMM-based potential calculation.

**Comment by jchodera** [comment]
> Question: how arbitrary is the ordering of the atoms in a topological torsion (1-2-3-4 vs 4-3-2-1)? I realize this probably shifts the angle, but since the potentials are calculated with their own angle (in the correct order for that torsion), should it make a difference when finding topological torsions?

The torsion angle computed from 1-2-3-4 should be identical to the one computed from 4-3-2-1.

**Comment by jchodera** [comment]
> This is probably a good time to move to OpenMM-based potential calculation.

+1 for that.  It would allow us to support CHARMM and other potentials.

**Comment by pgrinaway** [comment]
Thanks. I'll implement that stuff tonight then.

**Comment by pgrinaway** [comment]
Going to sleep. Still need to finish the torsion stuff, but most of it is done, and a bunch of duplicate code has been removed.

**Comment by pgrinaway** [comment]
Ok. Looks like the issue that I'm having right now is with the OpenMM context. When I try to instantiate one, it complains:
```pytb
  File "/Users/grinawap/PycharmProjects/perses/perses/rjmc/geometry.py", line 96, in propose
    logp_proposal, new_positions = self._logp_propose(top_proposal, current_positions, beta, direction='forward')
  File "/Users/grinawap/PycharmProjects/perses/perses/rjmc/geometry.py", line 184, in _logp_propose
    context = openmm.Context(growth_system, integrator, platform)
  File "/Users/grinawap/anaconda/envs/openeye-works/lib/python3.4/site-packages/simtk/openmm/openmm.py", line 6250, in __init__
    this = _openmm.new_Context(*args)
Exception: Unknown variable in expression: growth_stage
```

despite the fact that I have added the global parameter to each of the forces. I assume I'm missing something, but I'm not sure what. [This](https://github.com/choderalab/perses/pull/127/files#diff-a5d8d93e4a23e84d32747673e1244e75R659) is the method that creates the system.

**Comment by pgrinaway** [comment]
Nevermind, found my mistake.

**Comment by pgrinaway** [comment]
Debugging some other stuff

File: perses/rjmc/geometry.py [hunk 1]
```diff
@@ -582,162 +504,402 @@ def _propose_angle(self, angle, beta):
...
+        modified_angle_force.addPerAngleParameter("K")
+        modified_angle_force.addPerAngleParameter("growth_idx")
+        modified_angle_force.addGlobalParameter(parameter_name, 0)
+
+        modified_torsion_force = openmm.CustomTorsionForce(self._PeriodicTorsionForceEnergy.format(parameter_name))
+        modified_torsion_force.addPerTorsionParameter("periodicity")
+        modified_torsion_force.addPerTorsionParameter("phase")
+        modified_torsion_force.addPerTorsionParameter("k")
+        modified_torsion_force.addPerTorsionParameter("growth_idx")
+        modified_angle_force.addGlobalParameter(parameter_name, 0)
```
Comment (jchodera) [inline perses/rjmc/geometry.py]: This should be `modified_torsion_force`

**Comment by pgrinaway** [comment]
Currently an issue retrieving constraints...

**Comment by pgrinaway** [comment]
I have another angle issue...

**Comment by pgrinaway** [comment]
I wonder if there should be multiple torsion forces on the same atom set. Is this how OpenMM deals with multiplicity, or does this sound like a bug somewhere?

**Comment by pgrinaway** [comment]
Honestly not sure what's going on here. I've attached a plot of p(phi) for adding a methyl to ethane.

Dark blue - first carbon
Dark green - first hydrogen
Red - second hydrogen
teal - third hydrogen.

Kind of looks as I'd expect. However, the angles on the methyl are off (~127deg instead of ~108). The forces seem to be correct, including their growth stage as well.

adding_methyl

**Comment by pgrinaway** [comment]
Found the problem. I was turning forces off instead of on.

**Comment by jchodera** [comment]
Ooh!

**Comment by pgrinaway** [comment]
Have a weird issue retrieving constraints right now. Looking into it. Hopefully then we will be near the end of this saga...

**Comment by pgrinaway** [comment]
On the constraint front, it looks like the remaining weirdness is that there are C-H pairs where there is neither a bond term nor a constraint (there should be a constraint). This is in the system that `GeometryEngine` accepts as input, so I wonder if there is a problem building that system (propane...). I'm playing around with it now.

**Comment by pgrinaway** [comment]
Changed API. Fixed.

**Comment by pgrinaway** [comment]
Hm, the torsion scan potentials can still get `NaN`dy for transformations such as benzene-biphenyl (virtually all the positions have `NaN` potential). If the angles are excluded (by not adding them to the `growth_system`), then the `NaN`s go away, suggesting they come from the angle force. This is a bit disturbing. I need to prepare for tomorrow but I will continue looking into this.

**Comment by jchodera** [comment]
Can we merge this now (to at least fix the issues with `no eligible torsions found`) and continue on from there?

**Comment by jchodera** [comment]
Also the result for `benzene -> biphenyl` sounds totally reasonable to me.  This molecule is essentially completely planar, so most torsion options should be rejected.  If the concern is how we get adequate sampling of the small region of torsion space that has significant probability, we can certainly come up with a more clever way of doing that, but for now, a fine torsion discretization should do the trick.

**Comment by pgrinaway** [comment]
Ok, in that case, we can merge.

**Comment by pgrinaway** [comment]
Actually, I'm gonna commit one more thing.

**Comment by pgrinaway** [comment]
Ok, merging.

**Pull request merged**